examples/alignment.dx

```diff
@@ -0,0 +1,381 @@
+            True ->
+               j = ((ordinal i) * stride) @ m
+               k = ((ordinal i) * stride + window_size) @ m
+               x = for l':(j..k).
+                  l = %inject l'
+                  ls.l
+               unsafeCastTable window x
+ 
+            False -> zero
+
+' Add arbitrary padding based on type.
+
+def pad [Add out] (tab: m => out) : (n => out) =
+    for i.
+        case ordinal i < size m of
+             True -> tab.((ordinal i)@_)
+             False -> zero
+
+
+' ### Step 0: Preemphasis Smoothing
+
+' As a warmup, we will play around with windowing by running a simple
+convolution filter over the data. This preemphasis filter smooths out the signal.
+
+-- Hyperparam
+smoothing_coef = 0.97
+def smoothing (x:Fin 2=>Float)  : Float = x.(1@_) - (smoothing_coef * x.(0@_))
+    
+signal : Datsize => Float = map smoothing $ window 1 (for i. IToF raw_sample_wav.i)
+
+
+' ### Step 1: Movement to Frequency domain
+
+' Next we move from time to frequency domain by applying an FFT.
+
+' Instead of applying on FFT on the whole sequence we break it up into overlapping windows and apply an FFT on each one.  The windows in speech are known as frames, they are typically 0.025 seconds long. We use a stride of 0.01 seconds for overlap.
+
+-- Hyperparam
+WindowSize = Fin (FToI (0.025 * samplerate))
+stride = (FToI (0.01 * samplerate))
+
+
+PostWindow = Fin (idiv (size Datsize) stride)
+
+frame_split : PostWindow => WindowSize => Float = window stride signal
+
+
+' One issue though is that the FFT does not change the size of its input,
+  so if we have a short sequence we will end up with low frequency space
+  resolution. We get around this by padding out the time sequence.
+
+NFFT = Fin 512
+
+' Finally we only keep the positive components of the FFT which are the first half
+  plus one.
+
+Positive = Fin ((idiv 512  2) + 1)
+
+fft_frames = for i.
+    prefft: NFFT => Float = pad frame_split.i
+    ff = fft_real prefft
+    for j:Positive. (complex_abs ff.((ordinal j)@NFFT)) / (IToF (size NFFT))
+
+
+:html matshow fft_frames
+
+
+' ## Step 2 / 3: Construct Scales
+
+' The next problem is that we have the data in frequency-space which is not that useful for the
+  problem of speech recognition. We would like to move the data to a scale that is more focused on
+  the range of human communication and our biology.
+
+
+' ### Mel FilterBank
+
+' To do this we use a nonlinear rescaling known as the Mel Scale. 
+
+' Mel scale looks like this.
+
+Hertz = Float
+Mel = Float
+def mel (f:Hertz) : Mel = 1125.0 * log (1.0 + (f / 700.0))
+def imel (f:Mel) : Hertz = 700.0 * (exp (f / 1125.0) - 1.0)
+
+' However it is now so easy to move between frequency scales. To do this we will need
+' a bit of Dex machinery.
+
+' ### Linear Scales
+
+' Let us define a scaled range as a linear mapping between ordinals and a range. 
+
+LinearScale = {start:Float & end:Float}
+data ScaledRange a = AsScaledRange scale:LinearScale
```

Comment:
Many of the functions below need to know the `size` of a. (for instance `bin` and `step`) I thought it was cute to avoid passing and storing that by including the type. Do you think that is overkill?

Reply:
Ohhh nice. No I think this is fine.